x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/CachingUsernamePasswordRealm.java

```diff
@@ -78,74 +85,95 @@ public final void authenticate(AuthenticationToken authToken, ActionListener<Aut
-                            logger.debug("realm [{}] authenticated user [{}] (enabled:{}), with roles [{}]", name(), token.principal(),
-                                   user.enabled(), user.roles());
-                        }
-                        listener.onResponse(result);
-                    }, listener::onFailure));
+        try {
+            final SetOnce<User> authenticatedUser = new SetOnce<>();
+            final AtomicBoolean createdAndStartedFuture = new AtomicBoolean(false);
+            final ListenableFuture<Tuple<AuthenticationResult, UserWithHash>> future = cache.computeIfAbsent(token.principal(), k -> {
+                final ListenableFuture<Tuple<AuthenticationResult, UserWithHash>> created = new ListenableFuture<>();
+                if (createdAndStartedFuture.compareAndSet(false, true) == false) {
+                    throw new IllegalStateException("something else already started this. how?");
                 }
-            } else {
-                cache.invalidate(token.principal());
-                doAuthenticateAndCache(token, ActionListener.wrap((result) -> {
+                return created;
+            });
+
+            if (createdAndStartedFuture.get()) {
+                doAuthenticate(token, ActionListener.wrap(result -> {
                     if (result.isAuthenticated()) {
                         final User user = result.getUser();
-                        logger.debug("cached user's password changed. realm [{}] authenticated user [{}], with roles [{}]",
-                                name(), token.principal(), user.roles());
+                        authenticatedUser.set(user);
+                        final UserWithHash userWithHash = new UserWithHash(user, token.credentials(), hasher);
+                        future.onResponse(new Tuple<>(result, userWithHash));
+                    } else {
+                        future.onResponse(new Tuple<>(result, null));
                     }
-                    listener.onResponse(result);
-                }, listener::onFailure));
+                }, future::onFailure));
             }
-        } else {
-            cache.invalidate(token.principal());
-            doAuthenticateAndCache(token, ActionListener.wrap((result) -> {
-                if (result.isAuthenticated()) {
-                    final User user = result.getUser();
-                    logger.debug("cached user came from a lookup and could not be used for authentication. " +
-                            "realm [{}] authenticated user [{}] with roles [{}]", name(), token.principal(), user.roles());
+
+            future.addListener(ActionListener.wrap(tuple -> {
+                if (tuple != null) {
+                    final UserWithHash userWithHash = tuple.v2();
+                    final boolean performedAuthentication = createdAndStartedFuture.get() && userWithHash != null &&
+                        tuple.v2().user == authenticatedUser.get();
```

Comment:
@jaymode I think `tuple.v2().user == authenticatedUser.get()` is redundant (or there's a deeper reasoning that eludes me). Can you please clarify why this check is necessary?

Reply:
@albertzaharovits

This was introduced in response to my feedback above

- https://github.com/elastic/elasticsearch/pull/30794#discussion_r190140798

Reply:
That link doesn't seem to work correctly.
It's the 3rd comment in my review above.

Reply:
Thanks @tvernum . I understand, it's always better to test for the actual thing, if you can, instead of a surrogate. 
But as it makes you feel surer on the whole it makes me feel less of. The whole seems too gnarly: `handleResult` has both `createdAndStartedFuture` and `performedAuthentication` as params.
Then, I think we can do without `createdAndStartedFuture`.
I will open a nimble refactor non-issue on this.

Reply:
@tvernum I am going to concede on this one, for now 😭 .
I could not find a satisfactory refactoring such that the current behavior is maintained over all branches. Specifically I was looking to minimize `authenticationResult.isAuthenticated()` checks, as well as to split the listener into two types, one for the thread going for the network call and the other for the stalled threads. In the end I got a huge branching monster no better that the present one.
😢